simulation/src/main/scala/coop/rchain/caspersimulation/onchainstate/Resource.scala

```diff
@@ -2,65 +2,64 @@ package coop.rchain.caspersimulation.onchainstate
 
 import coop.rchain.caspersimulation.identity.{IdFactory, Identifiable}
 
+import scala.collection.immutable.HashSet
 import scala.util.Random
 
-sealed abstract class Resource extends Identifiable {
-  def isStopped: Boolean
-  def isProduce: Boolean
-  def isConsume: Boolean
-}
+sealed abstract class Resource extends Identifiable {}
 
 object Resource {
   private val rnd = new Random()
-  private val maxIndex = 15
 
   val producePrefix: String = "Produce-"
   val consumePrefix: String = "Consume-"
 
-  def matches(p: Produce, c: Consume): Boolean =
+  def consumes(resource: HashSet[Resource]): HashSet[Consume] = toConsumesProducesPair(resource)._1
+  def produces(resource: HashSet[Resource]): HashSet[Produce] = toConsumesProducesPair(resource)._2
+
+  private def toConsumesProducesPair(resource: HashSet[Resource]): (HashSet[Consume], HashSet[Produce]) =
+    resource.foldLeft(HashSet[Consume](), HashSet[Produce]()) {
+      case ((cs, ps), c: Consume) => (cs + c, ps)
+      case ((cs, ps), p: Produce) => (cs, ps + p)
+      case ((cs, ps), _) => (cs, ps)
+    }
+
+  def matches(c: Consume, p: Produce): Boolean =
     p.id.drop(producePrefix.length) == c.id.drop(consumePrefix.length)
 
   def random: Resource = {
     val isProduction = rnd.nextBoolean()
-    val index = rnd.nextInt(maxIndex).formatted("%04d")
+    // An arbitrary exponential function that puts 50% of the transactions to the first 33 indices
+    // and 99% of the transactions to the first 9766 indices.
+    // This is trying to roughly model the distribution of transactions
+    // across names in a production setting.
+    val exponentialRandom = (Math.pow(Math.log(rnd.nextDouble()), 3) / -0.01).toInt
+    val index = exponentialRandom.formatted("%04d")
     if (isProduction) {
       val continue = rnd.nextBoolean()
       if (continue) {
-        Produce(producePrefix + index, random)
+        Consume(producePrefix + index, random)
       } else {
-        Produce(producePrefix + index, Stopped)
+        Consume(producePrefix + index, Stopped)
       }
     } else {
-      Consume(consumePrefix + index)
+      Produce(consumePrefix + index)
     }
   }
 }
 
 case object Stopped extends Resource {
   override val id: String = "Stopped"
-
-  override def isStopped: Boolean = true
-  override def isProduce: Boolean = false
-  override def isConsume: Boolean = false
 }
 
-case class Produce(id: String, continuation: Resource = Stopped) extends Resource {
-  override def isStopped: Boolean = false
-  override def isProduce: Boolean = true
-  override def isConsume: Boolean = false
-}
+case class Consume(id: String, continuation: Resource = Stopped) extends Resource
 
-object Produce {
-  def apply(continuation: Resource)(implicit idf: IdFactory): Produce =
-    Produce(idf.next(Resource.producePrefix), continuation)
+object Consume {
```

Comment:
We should probably delete this companion object actually. I had created it initially because that was the pattern I was using with everything that extended `Identifiable`. But we only ever create these things through `Resource.random`, never through companion object factory methods.